### PR TITLE
Update zsh-functions

### PR DESCRIPTION
**Pull request by christiantakle**
I'm watching your [youtube video](https://www.youtube.com/watch?v=bTLYiNvRIVI) and it seems odd to me that you don't recurse in the else branch of `zsh_add_plugin`.

Logic: If the plugin doesn't exist *git clone* and then try again.

I kept the code in your style I normally run things through https://www.shellcheck.net/ or its cli.

cheers.

note/update:
For more safety you could add an attempt_count or something like 

```zsh
function zsh_add_plugin() {
    local -r plugin_name=$(basename "${1}")
    local -r plugin_dir="${ZDOTDIR}/plugins/${plugin_name}"
    local -r git_url="https://github.com/${1}.git"

    if [ -d "${plugin_dir}" ]; then 
        zsh_add_file "plugins/${plugin_name}/${plugin_name}.plugin.zsh" || \
        zsh_add_file "plugins/${plugin_name}/${plugin_name}.zsh"
    else
        if git clone "${git_url}" "${plugin_dir}"; then
            zsh_add_plugin "${1}"
        else
            echo "failed to git clone ${git_url}"
        fi
    fi
}
```

using lowercase variable names since they are local and readonly and keep global variables as UPPERCASE (YMMV).

using `basename` instead of the echo+cut. One could also do
```zsh
cut -d "/" -f 2 <<< "${1}"
```
Mostly pointless just showing some diversity

### take 2
No reason to have `zsh_add_file` depend on a global variable. We can change its to a general purpose `source_if_exist`.

```zsh
function source_if_exist() {
    [ -f "${1:?'provide: <file_path>'}" ] && source "${1}"
}

function zsh_add_plugin() {
    local -r plugin_name=$(basename "${1}")
    local -r plugin_dir="${ZDOTDIR}/plugins/${plugin_name}"
    local -r git_url="https://github.com/${1}.git"
    
    if [ -d "${plugin_dir}" ]; then 
        source_if_exist "${plugin_dir}/${plugin_name}.plugin.zsh" || \
        source_if_exist "${plugin_dir}/${plugin_name}.zsh"
    else
        if git clone "${git_url}" "${plugin_dir}"; then
            zsh_add_plugin "${1}"
        else
            echo "failed to git clone ${git_url}"
        fi
    fi
}
```

### CODE REVIEW

**Comment by christiantakle** [comment]
aha https://youtu.be/bTLYiNvRIVI?t=1399 :)

**Comment by bruhtus** [comment]
I want to ask, if you don't mind. In the else section of `zsh_add_plugin` function, why not make it from this:
```sh
else
        if git clone "${git_url}" "${plugin_dir}"; then
            zsh_add_plugin "${1}"
        else
            echo "failed to git clone ${git_url}"
        fi
fi
```
to this:
```sh
else
        git clone "${git_url}" "${plugin_dir}" && zsh_add_plugin "${1}" 2>/dev/null || echo "failed to git clone ${git_url}"
fi
```

Is there some reason for using an `if-else` statement rather than using `&&` and `||` in that case?
I'm still learning about shell scripting so I'm not sure whether to use `if-else` statement or `&&` and `||`. Thank you.

**Comment by christiantakle** [comment]
@bruhtus depending on the [precedence of operators](https://unix.stackexchange.com/questions/88850/precedence-of-the-shell-logical-operators) in the language they are equal.

Originally I had an example similar to yours but I would argue that its less readable to newcomer's.

You are also adding more constructs like `2> /dev/null` so know what that means. That is knowledge of stdin,stderr and shell redirection.

In this case it is more about if one prefer **statements** over **expressions**. I really like expression based languages as I primarily focus on functional programming but not for the sake of aesthetics or `just because I can`.

Its similar to [tacit programming](https://en.wikipedia.org/wiki/Tacit_programming#Functional_programming) (point-free or point-full style) when I was learning [haskell](https://www.haskell.org/) I obsessed about making everything point-free but as it turns out sometimes it doesn't matter or straight up just make code harder to read.

Hope it helps.

**Comment by bruhtus** [comment]
@christiantakle thank you for answering my question. Using `&&` and `||` is indeed less readable and sometimes I feel like it differs from `if-else` statements, here's an example:
```sh
# my plugin directory is on ~/.cache/zsh/plugins
[ -f "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh" ] && source "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh" || \
[ -f "$plugin_dir/$plugin_name/$plugin_name.zsh" ] && source "$plugin_dir/$plugin_name/$plugin_name.zsh" || \
[ -f "$plugin_dir/$plugin_name/$plugin_name.zsh-theme" ] && source "$plugin_dir/$plugin_name/$plugin_name.zsh-theme"
```
I'm trying to make something similar without `zsh_add_file` function using `&&` and `||` but for some reason the plugin keep going to the next expression. Let's say there's `something.plugin.zsh` instead of `something.zsh` file and, as far as I know, it should stop at the first expression but for some reason it keeps going to the second expression and causing an error like:
```sh
source:7: no such file or directory: /home/username/.cache/zsh/plugins/something/something.zsh
```

Do you know why that happen?
Maybe there's something wrong with my implementations?

Thank you.

**Comment by christiantakle** [comment]
It can be hard to follow the logical group when you chain so many logical operators together. It might be easier for you to follow what happends if you tried to write it as `if-else` statements.

You can try to add `{}` to make the grouping more clear e.g.

```
{ true && echo "a" } || { true && echo "b" } || { true && echo "c" }
a

{ false && echo "a" } || { true && echo "b" } || { true && echo "c" }
b

{ false && echo "a" } || { false && echo "b" } || { true && echo "c" }
c

{ true && false } || { true && echo "b" } || { true && echo "c" }
b
``` 
Three groupings where I show some of the results from changing some of the switches.

**Comment by bruhtus** [comment]
> ### Update:
> added `;`
> 
> What you are trying to do is quite nested
> 
> ```shell
> {
>     [ -f "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh" ] && source "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh"
> } || {
>     [ -f "$plugin_dir/$plugin_name/$plugin_name.zsh" ] && source "$plugin_dir/$plugin_name/$plugin_name.zsh"
> } || {
>     [ -f "$plugin_dir/$plugin_name/$plugin_name.zsh-theme" ] && source "$plugin_dir/$plugin_name/$plugin_name.zsh-theme"
> }
> ```
> 
> ```shell
> if [ -f "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh" ]; then
>     source "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh"
> elif [ -f "$plugin_dir/$plugin_name/$plugin_name.plugin.zsh" ]; then
>     source "$plugin_dir/$plugin_name/$plugin_name.zsh"
> elif [ -f "$plugin_dir/$plugin_name/$plugin_name.zsh-theme" ]; then
>     source "$plugin_dir/$plugin_name/$plugin_name.zsh-theme"
> fi
> ```
> 
> ### Update 2:
> Here is a quick test I did seem to work; using zsh (bash might want the `;`)
> 
> ```
> ➜  /tmp ls x.sh
> x.sh
> ➜  /tmp [ -f ./x.sh ] && echo a || [ -f ./x.sh ] && echo a
> a
> a
> ➜  /tmp {[ -f ./x.sh ] && echo a } || { [ -f ./x.sh ] && echo a }
> a
> ```
> 
> Apologize for the bad variable naming. You should be able to make a more complete example with a,b,c using the above if needed.

Oh, so I need to use `{}` for the nested expression. I thought I could do it like usual, and that's why I feel like `&&` and `||` not the same as `if-else` statement because I can use `elif`.

Alright, I learn something new today. Thank you!

**Comment by Abhishek416** [comment]
@ Christian Takle  i want  to ask you, 
in function source_if_exist()  
why do we want to call this   ```zsh_add_plugin "${1}"```  again 

```
if git clone "${git_url}" "${plugin_dir}"; then
            zsh_add_plugin "${1}"
```
kind of recursion technique
we can do this 

``` if [ ! git clone "${git_url}" "${plugin_dir}"]; then echo "failed to git clone ${git_url}" ```

**Comment by christiantakle** [comment]
@Abhishek416 Sorry havn't looked at this for a while so first see your message now.

Its because I want the recursive behaviour; If the plugin exist source it; otherwise try to clone and recurse; unless we failed to close the plugin.

Your code would not do the same.

**Comment by christiantakle** [comment]
I don't think my code will be accepted so should probably close it.

**Comment by christiantakle** [comment]
Closing due to lack of interest.
